### DISCUSSION TITLE
Magnetic Field Drifting

### DISCUSSION THREAD

**Discussion by nosoftwareskills** [tronikos/esphome-magnetometer-water-gas-meter#40] [Q&A]
My sensor is mounted with double sided tape as I test this out. Once I get good readings I will find a better way to secure it.

What I am seeing is the magnetic field for x,y, and z changes through the day. The difference between max and min is mostly consistent, but the average between them goes up and down. If I hit calibrate everything is good for about 10 minutes then I need to hit calibrate again.

I've tried a couple different sensors (from the same order) and it's the same thing. I haven't seen any similar comments. Anyone have thoughts what this could be? Does the sensor need calibration? Maybe these are bad sensors? TIA
![Screenshot_20241023-213822](https://github.com/user-attachments/assets/d129bbc1-3626-49ed-88f2-c8366623e11f)

**Comment by andrewperkov** [comment]
Have the same issue - readings are drifting, but will drop back to original eventually, then drifting again... Tried 3 or 4 different sensors. Suspecting a broken/not compatible water meter, ABB C700 in my case, which looks like this 
<img width="966" height="693" alt="image" src="https://github.com/user-attachments/assets/9206e410-075d-4315-94f1-13fe5c5f33bc" />

**Reply by eschmidt12** [reply]
I have the exact same issue. I have one magnetometer on my water meter that never drifts, but the magnetometer on my gas meter drifts like crazy

**Comment by pixelgrb** [comment]
I created a new thread (this one is more than a year old) .. I think my issue has to do with temperature: https://github.com/tronikos/esphome-magnetometer-water-gas-meter/discussions/85

**Reply by tronikos** [reply]
Try the latest version. Select the adaptive detection algorithm to handle temperature drift automatically

**Comment by pixelgrb** [comment]
@nosoftwareskills, @eschmidt12, @andrewperkov - can you share what sensors you have used?

I am experiencing the same issue with an Adafruit MMC5603.  The problem is related to temperature.  See: https://github.com/tronikos/esphome-magnetometer-water-gas-meter/discussions/85

I'm thinking of trying different sensors and would like to avoid the ones you already tried.

**Reply by andrewperkov** [reply]
I'm using https://www.aliexpress.com/item/1005008999430045.html   GY-271 QMC5883L,  which is actually qmc5883p. In my case, its is not a temperature issue I'm suspecting the water meter malfunction 
<img width="742" height="462" alt="image" src="https://github.com/user-attachments/assets/ef6f9b91-076d-49ea-83fc-89675f9003ec" />
.

**Reply by nosoftwareskills** [reply]
I am using an HMC5883L. But I don't think that sensor, nor this app is the problem.

My water meter is an AMCO C700, very similar to the one shared by @andrewperkov. There is an inside meter and two wires that go to an outside display. I am not positive about this but I'm guessing that display has two magnets in it, one for the inside meter and one to trigger a reed switch to cause the outside display to rotate. These two magnets moving at different speeds would create a vector sum of the electric fields making it harder to track.  I suspect the field would look something like the image shared by @andrewperkov and what I was also seeing.

My thinking that for this display it would be easier for me to read the pulses coming from the meter wires than track the magnetic field. The only concern is that the resolution is decreased significantly to once every 10 cubic feet (74.8 gallons) of water in my case.

**Reply by andrewperkov** [reply]
I do have exactly the same ABB C700 AKA AMCO/KENT C700 water meter.
